src/Compilers/CSharp/Portable/BoundTree/DecisionTree.cs

```diff
@@ -0,0 +1,916 @@
+            {
+                // We are switching on a constant expression. See if the constant value matches
+                bool matches = byValue.Expression.ConstantValue == value.ConstantValue;
+                if (!matches) return null;
+                byValue.Default = makeDecision(byValue.Expression, byValue.Type);
+                byValue.Default.Parent = byValue;
+                if (byValue.Default.MatchIsComplete)
+                {
+                    byValue.MatchIsComplete = true;
+                }
+                return byValue.Default;
+            }
+
+            // For error recovery, to avoid "unreachable code" diagnostics when there is a bad case
+            // label, we use the case label itself as the value key.
+            object valueKey = hasErrors ? (object)value : value.ConstantValue.Value;
+            DecisionTree valueDecision;
+            if (byValue.ValueAndDecision.TryGetValue(valueKey, out valueDecision))
+            {
+                return Add(valueDecision, makeDecision);
+            }
+            valueDecision = makeDecision(byValue.Expression, byValue.Type);
+            valueDecision.Parent = byValue;
+            byValue.ValueAndDecision.Add(valueKey, valueDecision);
+            return valueDecision;
+        }
+
+        private DecisionTree AddByValue(DecisionTree.ByType byType, BoundExpression value, DecisionMaker makeDecision, bool hasErrors)
+        {
+            if (byType.Default != null)
+            {
+                try
+                {
+                    return AddByValue(byType.Default, value, makeDecision, hasErrors);
+                }
+                finally
+                {
+                    if (byType.Default.MatchIsComplete) byType.MatchIsComplete = true;
+                }
+            }
+            if (value.ConstantValue == ConstantValue.Null)
+            {
+                return AddByNull((DecisionTree)byType, makeDecision);
+            }
+
+            foreach (var kvp in byType.TypeAndDecision)
+            {
+                var matchedType = kvp.Key;
+                var decision = kvp.Value;
+
+                // See if the test is already subsumed
+                switch (_conversions.ExpressionOfTypeMatchesPatternType(value.Type, matchedType))
+                {
+                    case true:
+                        if (decision.MatchIsComplete) return null;
+                        continue;
+                    case false:
+                    case null:
+                        continue;
+                }
+            }
+
+            DecisionTree forType = null;
+
+            // Find an existing decision tree for the expression's type
+            for (int i = byType.TypeAndDecision.Count - 1; i >= 0 && forType == null; i--)
+            {
+                var kvp = byType.TypeAndDecision[i];
+                var matchedType = kvp.Key;
+                var decision = kvp.Value;
+                if (matchedType == value.Type)
```

Comment:
Should this comparison consider dynamic, custom modifiers and array sizes differences?